templates/templates_test.go

```diff
@@ -89,3 +89,56 @@ func TestParseTruncateFunction(t *testing.T) {
 		})
 	}
 }
+
+func TestHeaderFunctions(t *testing.T) {
+	const source = "hello world"
+
+	tests := []struct {
+		doc      string
+		template string
+		expected string
+	}{
+		{
+			doc:      "json",
+			template: `{{ json .}}`,
+		},
+		{
+			doc:      "split",
+			template: `{{ split . ","}}`,
+		},
+		{
+			doc:      "join",
+			template: `{{ join . ","}}`,
+		},
+		{
+			doc:      "title",
+			template: `{{ title .}}`,
+		},
+		{
+			doc:      "lower",
+			template: `{{ lower .}}`,
+		},
+		{
+			doc:      "upper",
+			template: `{{ upper .}}`,
+		},
+		{
+			doc:      "truncate",
+			template: `{{ truncate . 2}}`,
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.doc, func(t *testing.T) {
+			tmpl, err := New("").Funcs(HeaderFunctions).Parse(tc.template)
+			assert.NilError(t, err)
+
+			var b bytes.Buffer
+			assert.NilError(t, tmpl.Execute(&b, source))
+
+			// All header-functions are currently stubs, and don't modify the input.
+			expected := source
```

Comment:
what's the purpose of  `expected` here?

Reply:
Mostly to make the output on failures less ambiguous; gotest.tools does not define an order in which "actual" and "expected" values must be passed (I _think_ this may have been on purpose, as there's been some bike-shedding about those, but I should really check with the maintainer if they would be open to have a recommendation added).


The `expected` intermediate variable is printed as part of the output, which helps reading "so, what was expected?" 😂 


Here's a quick example; it's most notably with the `DeepEqual` (showing a diff) where things are sometimes ambiguous ("which one is right?");

```go
func TestFoo(t *testing.T) {
	assert.Check(t, is.Equal("one", "two"))

	actual := "two"
	expected := "one"
	assert.Check(t, is.Equal(actual, expected))
	assert.Check(t, is.Equal(expected, actual))

	type foo struct {
		One, Two, Three string
	}

	assert.Check(t, is.DeepEqual(foo{One: "one", Two: "two", Three: "tree"}, foo{One: "one", Two: "not two", Three: "tree"}))

	expectedFoo := foo{One: "one", Two: "two", Three: "tree"}
	actualFoo := foo{One: "one", Two: "not two", Three: "tree"}
	assert.Check(t, is.DeepEqual(expectedFoo, actualFoo))
	assert.Check(t, is.DeepEqual(actualFoo, expectedFoo))
}
```

Which will output something like;


```console
=== RUN   TestFoo
    completions_test.go:34: assertion failed: one (string) != two (string)
    completions_test.go:38: assertion failed: two (actual string) != one (expected string)
    completions_test.go:39: assertion failed: one (expected string) != two (actual string)
    completions_test.go:45: assertion failed: 
        --- ←
        +++ →
          main.foo{
          	One:   "one",
        - 	Two:   "two",
        + 	Two:   "not two",
          	Three: "tree",
          }
        
    completions_test.go:49: assertion failed: 
        --- expectedFoo
        +++ actualFoo
          main.foo{
          	One:   "one",
        - 	Two:   "two",
        + 	Two:   "not two",
          	Three: "tree",
          }
        
    completions_test.go:50: assertion failed: 
        --- actualFoo
        +++ expectedFoo
          main.foo{
          	One:   "one",
        - 	Two:   "not two",
        + 	Two:   "two",
          	Three: "tree",
          }
        
--- FAIL: TestFoo (0.00s)
```

Reply:
oh i understand, thank you for the explanation. Was just a bit confused as it seemed like the assignment served no purpose :see_no_evil: